events/sinks/manager.go

```diff
@@ -0,0 +1,117 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sinks
+
+import (
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/heapster/events/core"
+)
+
+const (
+	DefaultSinkexportEventsTimeout = 20 * time.Second
+	DefaultSinkStopTimeout         = 60 * time.Second
+)
+
+type sinkHolder struct {
+	sink              core.EventSink
+	eventBatchChannel chan *core.EventBatch
+	stopChannel       chan bool
+}
+
+// Sink Manager - a special sink that distributes data to other sinks. It pushes data
+// only to these sinks that completed their previous exports. Data that could not be
+// pushed in the defined time is dropped and not retried.
+type sinkManager struct {
+	sinkHolders         []sinkHolder
+	exportEventsTimeout time.Duration
+	stopTimeout         time.Duration
+}
+
+func NewEventSinkManager(sinks []core.EventSink, exportEventsTimeout, stopTimeout time.Duration) (core.EventSink, error) {
+	sinkHolders := []sinkHolder{}
+	for _, sink := range sinks {
+		sh := sinkHolder{
+			sink:              sink,
+			eventBatchChannel: make(chan *core.EventBatch),
+			stopChannel:       make(chan bool),
+		}
+		sinkHolders = append(sinkHolders, sh)
+		go func(sh sinkHolder) {
+			for {
+				select {
+				case data := <-sh.eventBatchChannel:
+					sh.sink.ExportEvents(data)
+				case isStop := <-sh.stopChannel:
+					glog.V(2).Infof("Stop received: %s", sh.sink.Name())
+					if isStop {
+						sh.sink.Stop()
+						return
+					}
+				}
+			}
+		}(sh)
+	}
+	return &sinkManager{
+		sinkHolders:         sinkHolders,
+		exportEventsTimeout: exportEventsTimeout,
+		stopTimeout:         stopTimeout,
+	}, nil
+}
+
+// Guarantees that the export will complete in sinkexportEventsTimeout.
+func (this *sinkManager) ExportEvents(data *core.EventBatch) {
+	var wg sync.WaitGroup
+	for _, sh := range this.sinkHolders {
+		wg.Add(1)
+		go func(sh sinkHolder, wg *sync.WaitGroup) {
+			defer wg.Done()
+			glog.V(2).Infof("Pushing events to: %s", sh.sink.Name())
```

Comment:
It's probably a V(4) kind of log - I don't think this should be logged by default, especially because we also log in line 86.